cypress-io/cypress
# fix: Hovering over mount in command log does not show component in AUT

- Closes #24138 

### User facing changelog

Previously when hovering over `mount` in the reporter of a component test, the AUT shows a blank white view instead of the mounted component. This is not ideal if the user wants to see the component that has been mounted by the `mount` function.
This PR ensures that the snapshot of the mounted component is taken at the current time thereby guaranteeing component being displayed in the AUT

### Additional details

- Upon further investigation this issue only persists for the `vue2` and `react18` frameworks. For all other frameworks the intended behavior was happening i.e. The mount command shows the mounted component in the AUT.
- `React18` uses the `makeMountFn` from `react` under the hood therefore changes had to be made to the `mount` function inside `npm/react/createMount` for this framework.
- The proposed solution involves the use of the `.wait` and a `mutationObserver`. Using the `.wait(0, ...)` allows cypress to wait until the next tick in the event-loop to take a snapshot of the mounted component. This ensures that all the synchronous code (including the actual mounting of the component) has occurred before we attempt to take a snapshot. The `mutationObserver` listens to the target node (in our case the `data-cy-root`) to check if any element is added to it. If all the elements are added we can be sure that the component exists on the AUT and can thereby take the screenshot.

### Steps to test

The spec file `reporter-ct-mount-hover.cy.ts` iterates over several projects from `system-tests`, scaffolds, opens, and runs a spec file within each project while hovering on the mount function in the command log to ensure the intended functionality.

1. Go to the `packages/app`
2. Run `yarn cypress:open`
3. Select e2e testing on your preferred browser and navigate to the `reporter-ct-mount-hover.cy.ts` spec file.
4. This test iterates over all major frameworks testing hovering over the mount in the command-log ensuring that the component is visible in the AUT.

### How has the user experience changed?

Before: 
<img width="1080" alt="Screen Shot 2022-10-21 at 3 34 50 PM" src="https://user-images.githubusercontent.com/65267668/197275649-73d374e7-5b74-4f98-b7e2-ccc2edba6956.png">
After:
<img width="888" alt="Screen Shot 2022-10-21 at 3 29 19 PM" src="https://user-images.githubusercontent.com/65267668/197274776-762ef3df-a07e-4540-8f68-a7f50cc8f422.png">


### PR Tasks


- [na] Has the original issue (or this PR, if no issue exists) been tagged with a release in ZenHub? (user-facing changes only)
- [na] Has a PR for user-facing changes been opened in [`cypress-documentation`](https://github.com/cypress-io/cypress-documentation)? 
- [na] Have API changes been updated in the [`type definitions`](https://github.com/cypress-io/cypress/blob/develop/cli/types/cypress.d.ts)?
